externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp

```diff
@@ -111,4 +111,52 @@ LogicalResult IsolatedGroupOp::verify() {
 
 OpFoldResult ConstOp::fold(FoldAdaptor) { return getValueAttr(); }
 
+LogicalResult ConcatOp::verify() {
+  auto outputTensorType = getResult().getType().cast<TensorType>();
+  int64_t axis = getAxis();
+  int64_t concatDimVal = outputTensorType.getShape()[axis];
+  // Accumulate concat dim, ignored if any of the inputs is dynamic.
+  int64_t concatDimAcc = 0;
+  // All tensors must have the smae dtype, rank and all non-concat dims must be
+  // the same.
+  for (auto type : getInputs().getTypes()) {
+    auto inputTensorType = type.cast<TensorType>();
+    if (inputTensorType.getElementType() != outputTensorType.getElementType())
+      return emitOpError() << "failed to verify tcp.concat with input and "
+                              "output operands rank mismatch";
+    if (outputTensorType.getRank() != inputTensorType.getRank())
+      return emitOpError() << "failed to verify tcp.concat with input and "
+                              "output operands rank mismatch";
+    for (int64_t dim = 0; dim < inputTensorType.getRank(); ++dim) {
+      if (dim == axis) {
+        concatDimAcc += inputTensorType.getShape()[dim];
+      } else {
+        if (inputTensorType.getShape()[dim] != outputTensorType.getShape()[dim])
+          emitOpError() << "failed to verify tcp.concat with non concat dim "
+                        << dim << ", having different values "
+                        << inputTensorType.getShape()[dim] << " "
+                        << outputTensorType.getShape()[dim];
+      }
+    }
+  }
+
+  // If concat dim is dynamic, at least one input must be dynamic.
```

Comment:
This check is assuming that shape propagation has already happened before converting to TCP. Is that always guaranteed? 

I understand Torch does shape propagation. But what if there are other frontend dialects that convert to TCP but do not perform shape propagation completely?

In the case, would we end up with this scenario where the inputs are static and the output is dynamic (because its shape hasn't be propagated fully)?

Reply:
If we want to support type inference at the TCP level, we have to not use `?` or dynamic shape  to carry the semantics of both `dynamic` & `unknown` dim. I can see many issues including doing that.